### PR TITLE
feat(kubernetes platform): Expose node-port for vector-aggregator

### PR DESCRIPTION
**Pull request by langecode**
Allow for configuring the vector-aggregator to be exposed on a node port such that the aggregator can receive traffic from outside of the cluster where it is deployed.

### CODE REVIEW

**Comment by MOZGIII** [comment]
Thank you for this PR!

As I understand you want to be able to customize the Service parameters, but only for the service with an IP attached, and not for the headless one?
More specifically - you want to expose the service as a node port.

I understand why you need this.

Implementation wise though, I think we'd like to try a different approach - instead of replacing manual substitution of the `vector-aggregator.servicePorts`, I think we should improve the `vector-aggregator.servicePorts` define to allow generating both headless and normal service data (which are different with this approach). This will allow us to prevent the templates for the two from divering.
Alternatively, we could add `vector-aggregator.headlessServicePorts` in addition to `vector-aggregator.servicePorts`.

In addition to the concerns above, I've compiled a small to-do list of the trivial things to be done in this PR before it can be merged:
- [ ] the `values.yaml` examples for the new parameters
- [ ] run and commit the `make update-kubernetes-yaml` to make the check pass
- [ ] run and commit the `make update-helm-snapshots` to make the check pass

We'll be inspecting the changes to the generated files as part of this PR flow.

**Review by MOZGIII — CHANGES_REQUESTED** [review]
See my comment above:
- need a consensus on how the `distribution/helm/vector-aggregator/templates/service-headless.yaml` changes
- the to do list

**Comment by langecode** [comment]
I understand your motivation for the templates. Probably either both services should be templated or both should just be inlined. I see your argument with the risk of the service specifications drifting apart, however creating a template that is aware of the context in which it is being put to use (headless or not) is also not very appealing.

By the way, what is the use case for the headless service? I see it is being referenced from the statefulset to be used as base for the dns entries of the pods in the statefulset but are these entries necessary? I assume there is no communication between the Vector instances and it is not a requirement that a statefulset has a headless service associated.

I'll see if I can come up with a way to solve your concerns more elegantly.

**Comment by langecode** [comment]
Thank you for the feedback. I have changed the rendering a bit so that the ports array of the regular service is the basis for the headless service but for each port a possible `nodePort` property is removed - not only from the Vector source but also from remaining ports.

Just for information, the use case it running the aggregator in one Kubernetes cluster receiving data from other clusters. Hence, we need some way of doing ingress into the cluster. Since this is not pure http we would like to be able to do either a `NodePort` service or a `LoadBalancer` service - in the particular case there is no cloud controller installed capable of provisioning a load balancer so it has to be `NodePort`.

Btw, I ran the make tagets, that is the I could not find the `update-helm-snapshots` target. There is an `update-helm-dependencies` which I believe should not change anything.

**Comment by MOZGIII** [comment]
> it is not a requirement that a statefulset has a headless service associated.

It is a requirement for a StatefulSet afaik, that's why we're even bothering with it in the first place.

**Comment by langecode** [comment]
You are right. I thought is was optional - but it is not [StatefulSetSpec v1 apps](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#statefulsetspec-v1-apps). Out of curiosity what is the reasoning for doing the workload as statefulset instead of deployment? To have a stable persistent volume for the aggregators?

One other option could be to always create the headless service without any ports then.

**Comment by MOZGIII** [comment]
> To have a stable persistent volume for the aggregators?
Yes, it's inportnatn for vector have access to some storage to keep it's buffers.

> One other option could be to always create the headless service without any ports then.
There is some problem with that, as I recall. Some combination of options is invalid for the Kubernetes cluster, but it's only discoverable when the Service is submitted to the API and rejected. We should test all the various possible options.

Regarding that snapshot command - I guess it hasn't been merged yet. I think I'll merge it sooner rather than later, so we can write some extra tests for the functionality this PR adds.

---

I don't like how we ended up with different shapes of outputs from the templates, and also that they now need a `ports` inline. The first issue can be solved by extracing a "shared" part of the template into another template, and wiring the notmal service template via the same `toYaml`/`fromYaml` calls.

But other than that - I think we can use an extra argument to the template to signal whether to render as a notmal or headless service. This way we can avoid using `toYaml`/`fromYaml` and just use `if`s - as usual.

The trick to pass an extra artument is to use a `tuple` fn. You can find examples in the other parts of the chart (possible after we merged the helm snapshots PR, as I recall that PR was using them for sure, but we it might be the first time we used them).

**Comment by MOZGIII** [comment]
The PR I'm talking about: https://github.com/timberio/vector/pull/5673

**Comment by langecode** [comment]
The reasoning for doing the to/from yaml was to remove possible `nodePorts` from the user defines extra ports also. Because it would be the exact same problem if I wanted to add my own port definition with a `nodePort`. Otherwise I would not have gone through that iteration thing because I agree it is more complex to read.

If you want to parameterize the template that is of course one way to go. Imho, that just kind of gives the template knowledge about in which contexts it is used - I would much rather solve that with different templates.

**Comment by MOZGIII** [comment]
> Imho, that just kind of gives the template knowledge about in which contexts it is used - I would much rather solve that with different templates.

It does indeed, but ultimately this is the goal, so I think it is fine. To me, it feels like the direct approach - telling the template it's operational context - is the easier to maintain in the long run - that's why I chose it.

**Comment by MOZGIII** [comment]
> Btw, if it is not possible to deploy the service with an empty ports array I think there is actually another issue. Right now if the Vector source is disabled and no ports is added that would be what you get. I guess, that configuration might not make much sense but anyway - maybe a validation should be in place.

Currently the templates handle this situation correctly. I'm exploring ways to improve the validation of this specific section - but we'll likely need to dry-rin these against the real cluster.

**Comment by langecode** [comment]
> > Imho, that just kind of gives the template knowledge about in which contexts it is used - I would much rather solve that with different templates.
> 
> It does indeed, but ultimately this is the goal, so I think it is fine. To me, it feels like the direct approach - telling the template it's operational context - is the easier to maintain in the long run - that's why I chose it.

Yes, I kind of figured that would be the position of the project here :-)

While I do not agree I think I will see if I can find some time to change the PR. I think the structure of #6175 makes sense though - to assume the `service.ports` contains the necessary properties for a service port definition. This eliminates the need to all the `toYaml` and `fromYaml` to remove the `nodePort` property.

**Comment by langecode** [comment]
> > Btw, if it is not possible to deploy the service with an empty ports array I think there is actually another issue. Right now if the Vector source is disabled and no ports is added that would be what you get. I guess, that configuration might not make much sense but anyway - maybe a validation should be in place.
> 
> Currently the templates handle this situation correctly. I'm exploring ways to improve the validation of this specific section - but we'll likely need to dry-rin these against the real cluster.

What I meant was the below scenario - but probably a bit theoretical :-)

```
➜  vector git:(feature/allow-aggregator-nodeport-binding) helm template distribution/helm/vector-aggregator --set vectorSource.enabled=false
...
# Source: vector-aggregator/templates/service-headless.yaml
apiVersion: v1
kind: Service
metadata:
  name: RELEASE-NAME-vector-aggregator-headless
  labels:
    helm.sh/chart: vector-aggregator-0.0.0
    app.kubernetes.io/name: vector-aggregator
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "0.0.0"
    app.kubernetes.io/managed-by: Helm
spec:
  clusterIP: None
  ports:
    []
  selector:
    app.kubernetes.io/name: vector-aggregator
    app.kubernetes.io/instance: RELEASE-NAME
...
```

**Comment by MOZGIII** [comment]
The thing is, if we don't have any ports, the logic is quite complicated:
- we can't have a "normal" Service with empty `ports` or without `ports` - so we skip it entirely;
- the headless service though is required for StatefulSet, and it *is allowed* to have empty `ports` - so we must provide it even if the ports are empty.


Current code does this properly, however this nuance prenets us from going with a simple and elegant templating logc there. This is why I think having a template with conditional inputs is superior.

Thinking about this design more, I think we would like to provide more flexibility on how we customize the services - yet still without going into the wildcard `toYaml .Values.serviceparams` like unstructured approach.

From the outside, we should probably have three templates/"entrypoints" in the shared chart - for normal service ports, for the headless service ports and for whether we want to omit or generate the normal service as a whole (i.e. whether the contents of the ports from the shared chart is empty or not).

I beleive this calling interface captures the flexibility just right: it allows the template embedders to preserve full control over the generated data (i.e. it can mix in some specific ports if needed), while allowing the underlying template implementations to be built with just limited and simple invariants in mind. This is good, because both shared templates and the resulting services can then be easily iterated on, with minimal risk of breaking this structure (i.e. requiring a refactor). In other words, I think it will be a stable architecture in the long run - for this small bit.

Under the hood, both templates generating the ports should probably reuse some internal template (as we do for vector config). This will allow us to hide complexity of the tuple-based calling convention behind a traditional "simple `.`" pattern.

**Comment by langecode** [comment]
I see. I did not check the doc there on whether the ports array could be
empty. I just assumed from our talk here that the ports could not be empty.
But if the headless service is there only for the sake of the statefulset
then one other option (as mentioned in
https://github.com/timberio/vector/pull/6020#issuecomment-760165370) could
be to just omit the ports from the headless service always.

I think the idea of having three "entrypoints" is much better. The thing I
do not like with the tuple approach is the very implicit contract towards
the template. The caller does not have a parameter list. The caller
basically just has to know that this template takes a tuple and that the
template has different flavors depending on a kind of "enumeration"
parameter in the tuple. I much prefer such things to be explicit so the
caller chose by name instead of by sending in a parameter. Then having the
"entrypoint" template internally delegate by using the tuple pattern is
another thing (as if you had a "private" scope). I'll find some time to
implement this approach combined with the structure from #6175 :-)

>

**Comment by langecode** [comment]
Note, the construction of the service port properties is based on the current Kubernetes specification with respect to which properties are optional and also the recommendation that `targetPort` is omitted for headless service ([ServicePort v1 core](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#serviceport-v1-core)).

**Comment by MOZGIII** [comment]
> Note, the construction of the service port properties is based on the current Kubernetes specification with respect to which properties are optional and also the recommendation that `targetPort` is omitted for headless service ([ServicePort v1 core](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#serviceport-v1-core)).

Thanks, this looks much better! The next logical step I'd like to take since we're at it is extracting the template for the `ServicePort` into `vector-shared`, as it seems reusable across all of the charts. I'll also leave some comments on the code inline - but overall this looks good.

File: distribution/helm/vector-aggregator/templates/_helpers.tpl [hunk 1]
```diff
@@ -1,18 +1,55 @@
...
 */}}
-{{- define "vector-aggregator.servicePorts" -}}
-{{- if .Values.vectorSource.enabled }}
+{{- define "vector-aggregator.internalServicePorts" -}}
+{{- $headless := index . 0 -}}
+{{- $values := index . 1 -}}
+{{- if $values.vectorSource.enabled }}
 - name: vector
-  port: {{ .Values.vectorSource.listenPort }}
+{{- if and $values.vectorSource.nodePort (not $headless) (eq "NodePort" $values.service.type) }}
```
Comment (MOZGIII) [inline distribution/helm/vector-aggregator/templates/_helpers.tpl:12]:
```suggestion
{{- if and $values.vectorSource.nodePort (not $headless) }}
```

We should set the `nodePort` if it's defined, regardless of what's the service type is. This will work better for `LoadBalancer` type, that also respects the `nodePort` setting.

File: distribution/helm/vector-aggregator/templates/_helpers.tpl [hunk 2]
```diff
@@ -1,18 +1,55 @@
...
+{{- if not $headless }}
+  targetPort: {{ $values.vectorSource.listenPort }}
+{{- end }}
+{{- end }}
+{{- range $values.service.ports }}
+- port: {{ .port }}
+{{- if not $headless }}
+  targetPort: {{ .targetPort }}
+{{- end }}
+{{- if and .nodePort (not $headless) (eq "NodePort" $values.service.type) }}
```
Comment (MOZGIII) [inline distribution/helm/vector-aggregator/templates/_helpers.tpl:26]:
```suggestion
{{- if and .nodePort (not $headless) }}
```
See https://github.com/timberio/vector/pull/6020/files#r562479964

File: distribution/helm/vector-aggregator/templates/service-headless.yaml [hunk 3]
```diff
@@ -7,8 +7,6 @@ metadata:
 spec:
   clusterIP: None
   ports:
-    {{- if eq (include "vector-aggregator.servicePortsPresent" .) "true" -}}
```
Comment (MOZGIII) [inline distribution/helm/vector-aggregator/templates/service-headless.yaml:10]: This wrapper should remain to be consistent with the non-headless service.

File: distribution/helm/vector-aggregator/templates/_helpers.tpl [hunk 4]
```diff
@@ -1,18 +1,55 @@
...
+{{- with .appProtocol }}
+  appProtocol: {{.}}
 {{- end }}
-{{- with .Values.service.ports }}
-{{ toYaml . }}
 {{- end }}
+{{- end -}}
+
+{{/*
+Resolve effective service ports to use.
```
Comment (MOZGIII) [inline distribution/helm/vector-aggregator/templates/_helpers.tpl:42]:
```suggestion
Generate effective service ports for normal (non-headless) service definition.
```

File: distribution/helm/vector-aggregator/templates/_helpers.tpl [hunk 5]
```diff
@@ -1,18 +1,55 @@
...
+*/}}
+{{- define "vector-aggregator.servicePorts" -}}
+{{- tuple false .Values | include "vector-aggregator.internalServicePorts" -}}
+{{- end -}}
+
+{{/*
+Generate effective service ports omitting the 'nodePort' for headless definition.
```
Comment (MOZGIII) [inline distribution/helm/vector-aggregator/templates/_helpers.tpl:49]:
```suggestion
Generate effective service ports for headless service definition.
```

File: distribution/helm/vector-aggregator/values.yaml [hunk 6]
```diff
@@ -191,6 +193,8 @@ vectorSource:
     # option: "value"
   # Raw TOML config to embed at the vector source (deprecated).
   rawConfig: null
+  # Specific node port to bind if `service.type` is set to `NodePort`.
+  nodePort: 0
```
Comment (MOZGIII) [inline distribution/helm/vector-aggregator/values.yaml:197]:
```suggestion
  # Specific node port to bind (useful if `service.type` is set to `NodePort`).
```

File: distribution/helm/vector-aggregator/values.yaml [hunk 7]
```diff
@@ -191,6 +193,8 @@ vectorSource:
     # option: "value"
   # Raw TOML config to embed at the vector source (deprecated).
   rawConfig: null
+  # Specific node port to bind if `service.type` is set to `NodePort`.
```
Comment (MOZGIII) [inline distribution/helm/vector-aggregator/values.yaml:196]:
```suggestion
  # Specific node port to bind (useful if `service.type` is set to `NodePort`).
```

File: distribution/helm/vector-aggregator/values.yaml (hunk 6 above)
Comment (MOZGIII) [inline distribution/helm/vector-aggregator/values.yaml:197]:
```suggestion
  nodePort: null  # 9090 (must be a number)
```

Not sure how it would play together with the changes at https://github.com/timberio/vector/pull/6020/files#r562479964, so it might be useful to change like this.

**Review by MOZGIII — CHANGES_REQUESTED** [review]
https://github.com/timberio/vector/blob/e6659f9c015065436c0e885f127103d40931454e/distribution/helm/vector-aggregator/values.yaml#L153-L162

We should add `nodePort` here too.

File: distribution/helm/vector-aggregator/values.yaml (hunk 7 above)
Comment (MOZGIII) [reply distribution/helm/vector-aggregator/values.yaml:196]: Also, I'd prefer if the `rawConfig` remained the last field. Could we move the `nodePort` above it?

File: distribution/helm/vector-aggregator/templates/service-headless.yaml (hunk 3 above)
Comment (langecode) [reply distribution/helm/vector-aggregator/templates/service-headless.yaml:10]: Sure I can put it back in, however, it ends up the same output and removing a branch point reduces cyclomatic complexity.

Comment (MOZGIII) [reply distribution/helm/vector-aggregator/templates/service-headless.yaml:10]: I agree this is not pretty, but we should either use that in both places - merely for consistency - or figure out another way to do it.
What we could do, I guess, is instead of doing
```
{{- if eq (include "vector-aggregator.servicePortsPresent" .) "true" -}}
```
at the `vector/distribution/helm/vector-aggregator/templates/service.yaml`, try something like
```
{{- if not (empty (include "vector-aggregator.servicePorts" .)) -}}
```
and remove the `vector-aggregator.servicePortsPresent` itself.

**Comment by langecode** [comment]
> Thanks, this looks much better! The next logical step I'd like to take since we're at it is extracting the template for the `ServicePort` into `vector-shared`, as it seems reusable across all of the charts. I'll also leave some comments on the code inline - but overall this looks good.

You would like it to go into the shared chart as-is (of course changing the template naming prefix)?

**Comment by MOZGIII** [comment]
I was thinking about splitting out a template for an individual `ServicePort`, and moving that into the shared chart. Then resuing it from the shared chart at the `vector-aggregator.internalServicePorts` (which should remain at the `vector-aggregator` chart).

File: distribution/helm/vector-aggregator/templates/_helpers.tpl [hunk 8]
```diff
@@ -1,18 +1,39 @@
...
+{{- $headless := index . 0 -}}
+{{- $values := index . 1 -}}
+{{- if $values.vectorSource.enabled }}
+{{- $servicePort := merge (dict) $values.vectorSource -}}
+{{- $_ := set $servicePort "name" "vector" -}}
+{{- $_ := set $servicePort "port" $values.vectorSource.listenPort -}}
+{{- $_ := set $servicePort "protocol" "TCP" -}}
+{{- if not $headless -}}
+{{- $_ := set $servicePort "targetPort" $values.vectorSource.listenPort -}}
+{{- end -}}
```
Comment (MOZGIII) [inline distribution/helm/vector-aggregator/templates/_helpers.tpl:17]:
```suggestion
{{- $servicePort := dict -}}
{{- $_ := set $servicePort "name" "vector" -}}
{{- $_ := set $servicePort "port" $values.vectorSource.listenPort -}}
{{- $_ := set $servicePort "protocol" "TCP" -}}
{{- $_ := set $servicePort "targetPort" $values.vectorSource.listenPort -}}
```

I'm pretty sure we don't want to inherit the `$values.vectorSource` here, but to construct the service port explicitly. Also, no need to conditionally hide the target port for headless service, as it's handled at the `include "libvector.servicePort"` (which was the point for this change).

File: distribution/helm/vector-aggregator/values.yaml [hunk 9]
```diff
@@ -152,12 +152,15 @@ psp:
...
 service:
+  # Service type - defaults to `ClusterIP` (only relevant for the regular service).
+  type: "ClusterIP"
   # Additional ports to expose.
   ports: []
     # - name: http
     #   port: 1234
     #   protocol: TCP
     #   targetPort: 1234
+    #   nodePort: 1243  (optional but useful with `service.type` is set to `NodePort`)
```
Comment (MOZGIII) [inline distribution/helm/vector-aggregator/values.yaml:163]:
```suggestion
    #   nodePort: 1243  # (optional but useful with `service.type` is set to `NodePort`)
```

Comment (eeyun) [reply distribution/helm/vector-aggregator/values.yaml:163]: good  catch!

File: distribution/helm/vector-aggregator/templates/_helpers.tpl (hunk 8 above)
Comment (langecode) [reply distribution/helm/vector-aggregator/templates/_helpers.tpl:17]: Yeah, sure - I actually did discuss with myself there also :-) Just need to include the `nodePort` also as that is the only values that is inherited. I'll change that one.

**Review by MOZGIII — APPROVED** [review]
LGTM

**Comment by MOZGIII** [comment]
Thanks for your contribution!

**Comment by langecode** [comment]
You're welcome - I think we surely ended up with a nice solution :-)
